In [375]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import holoviews as hv

In [376]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [377]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [378]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [379]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit(df_market_data)
market_data_transformed = market_data_scaled.transform(df_market_data)

In [380]:
# Create a DataFrame with the scaled data
df_market_data_transformed = pd.DataFrame(market_data_transformed,
                                          columns=["price_change_percentage_24h",
                                                   "price_change_percentage_7d",
                                                   "price_change_percentage_14d",
                                                   "price_change_percentage_30d",
                                                   "price_change_percentage_60d",
                                                   "price_change_percentage_200d",
                                                   "price_change_percentage_1y"]
                                                   )

# Copy the crypto names from the original data
df_market_data_transformed['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_transformed = df_market_data_transformed.set_index('coin_id')

# Display sample data
df_market_data_transformed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [381]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [382]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_market_data_transformed)
    inertia.append(k_model.inertia_)

/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will cha

In [383]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k':k,
              'inertia':inertia
              }

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [384]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x='k',
    y='inertia',
    xticks=k,
    title="Elbow Plot"
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on the chart, the ideal value for `k` seems to be `4`.


---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [385]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [386]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_transformed)

/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4)

In [387]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k4 = model.predict(df_market_data_transformed)

# Print the resulting array of cluster values.
k4

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 2, 3, 1, 1, 0, 1, 1, 1, 1],
      dtype=int32)

In [388]:
# Create a copy of the DataFrame
df_predictions = df_market_data_transformed.copy()

In [389]:
# Add a new column to the DataFrame with the predicted clusters
df_predictions['k4']=k4

# Display sample data
df_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  k4  
coin_id                                       
bitcoin                        -0.251637   3  
ethereum                       -0.199352   3  
tether                         -0.282061   1  
ripple                         -0.295546   1  
bitcoin-cash                   -0.270317   3

In [390]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by='k4',
    title='Predictions for k4'
)

:NdOverlay   [k4]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In [391]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [392]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_transformed_pca = pca.fit(df_market_data_transformed)
market_data_transformed_pca = market_data_transformed_pca.transform(df_market_data_transformed)
# View the first five rows of the DataFrame. 
market_data_transformed_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [393]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is `0.3719856 + 0.34700813 + 0.17603793 = 0.895` 

In [394]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data
df_market_data_transformed_pca = pd.DataFrame(market_data_transformed_pca, columns=['PCA1','PCA2','PCA3'])

# Copy the crypto names from the original data
df_market_data_transformed_pca['coin_id'] = df_market_data_transformed.index

# Set the coinid column as index
df_market_data_transformed_pca = df_market_data_transformed_pca.set_index('coin_id')

# Display sample data
df_market_data_transformed_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [395]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [396]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i,random_state=1)
    k_model.fit(df_market_data_transformed_pca)
    inertia.append(k_model.inertia_)

/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will cha

In [397]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k':k,
              'inertia':inertia
              }
# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [398]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x='k',
    y='inertia',
    xticks=k,
    title='Elbow Plot'
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Based on the chart, the ideal value for `k` seems to be `4`.



* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** `No`, the k value is the same as found using the original data.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [399]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [400]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_transformed_pca)

/Users/joaopedrofortunato/opt/anaconda3/envs/PythonData/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4)

In [401]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k4_pca = model.predict(df_market_data_transformed_pca)
# Print the resulting array of cluster values.
k4_pca

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 2, 0, 1, 1, 3, 1, 1, 1, 1],
      dtype=int32)

In [402]:
# Create a copy of the DataFrame with the PCA data
df_predictions_pca = df_market_data_transformed_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_predictions_pca['k4_pca'] = k4_pca

# Display sample data
df_predictions_pca.head()

PCA1      PCA2      PCA3  k4_pca
coin_id                                           
bitcoin      -0.600667  0.842760  0.461595       0
ethereum     -0.458261  0.458466  0.952877       0
tether       -0.433070 -0.168126 -0.641752       1
ripple       -0.471835 -0.222660 -0.479053       1
bitcoin-cash -1.157800  2.041209  1.859715       0

In [403]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_predictions_pca.hvplot.scatter(
    x='PCA1',
    y='PCA2',
    by='k4_pca',
    title='Predictions for k4_pca'
)

:NdOverlay   [k4_pca]
   :Scatter   [PCA1]   (PCA2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [404]:
# Composite plot to contrast the clusters
composite_plot = (
    df_predictions.hvplot.scatter(
        x="price_change_percentage_24h",
        y="price_change_percentage_7d",
        by='k4',
        title='Predictions for k4',
        width=400,
        height=400
    )
    + df_predictions_pca.hvplot.scatter(
        x='PCA1',
        y='PCA2',
        by='k4_pca',
        title='Predictions for k4_pca',
        width=400,
        height=400
    )
).opts(shared_axes=True)

composite_plot


:Layout
   .NdOverlay.I  :NdOverlay   [k4]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .NdOverlay.II :NdOverlay   [k4_pca]
      :Scatter   [PCA1]   (PCA2)

In [405]:
hv.save(composite_plot, 'Outputs/composite_plot.png')


#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features in K-Means clustering saves time, aids visualization, and reduces noise. However, in this case in specific, seems like it may have led to information loss and lower cluster quality.